PlanningBiblio/PlanningBiblio
# Mt29957 (#753)

* MT29957: Password complexity

Asks the user to provide a robust enough password on password change.

The following rules are implemented:

 - the password must be at least 8 characters long.
 - the password must contain at least one digit.
 - the password must contain at least one capital letter.
 - the password must contain at least one lower-case letter.
 - the password must contain at least one special character, from the following list: !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ ( comes from https://owasp.org/www-community/password-special-characters )

Also, the code has been refactored so:

 - the code checking the complexity is the same on client side and server side validation.
 - the UI for the password change is the same in "My Account" and Agent edition.

Test plan:

 - Log in with a user with the following rights : Gestion des agents.
 - Check that you can change the password of any agent, only if the complexity requirements are met.
 - Check that you cannot change the password without the "Gestion des agents" right.
 - Check that you can change your own password on the "My Account" page, regardless of your rights.

* MT29957, Follow-up: Password complexity

 - Remove error messages and passwords on cancel
 - Add current password check when the user is changing its own password

* MT29957, Follow-up: Password complexity

 - Clean passwords and error messages from the modal on any type of close

* MT29957, Follow-up: Password complexity

Add a new configuration option : "Longueur-minimum-mot-de-passe" which allows to define the minimum number of characters for a password.

This value cannot be lower than 8.

* MT29957, Follow-up: Password complexity

Change the parameter name to "Auth-PasswordLength"

* MT29957, Follow-up: Password complexity

Add CSRF protection to routes ajax/change-own-password and ajax/change-password

* MT29957, Follow-up: Password complexity

 - Fix login change
 - Add title in My Account tab

* MT29957, Follow-up: Password complexity

 - Make sure change_own_password is always initialised.

* MT29957, Follow-up: Password complexity

 - Restore public/personnel/password.php, which is still needed.

* MT29957, Follow-up: Password complexity

 - Use new method instead of public/personnel/password.php

Known limitation: menu password change is disabled on agent edition (because there already is a password change on this page)

* MT29957, Follow-up: Password complexity

 - Remove the ability to change an agent's password on agent edition page.

* MT29957, Follow-up: Password complexity

 - Don't use a pop-up for password change
 - Display a confirmation message when the password has ben successfully updated

* MT29957, Follow-up: Password complexity

 - Update Cancel button style
 - Update link for non-symfonised menu

* MT29957, Follow-up: Password complexity

 - Allow to see password input on password change

* MT29957, Follow-up: Password complexity

- Change routes ajax.checkpassword and ajax.iscurrentpassword from POST to GET
- Close the label show_passwords_checkbox

Co-authored-by: Matthias Meusburger <matthias.meusburger@biblibre.com>
Co-authored-by: Jérôme Combes <jerome@planningbiblio.fr>
